### ISSUE TITLE
elasticdl.layers.Embedding perf

### ISSUE THREAD

**Issue by skydoorkai** [sql-machine-learning/elasticdl#1084]
Tested on 8/27/2019 for deepFM model.

model | total time | compute gradient time | lookup embedding time in compute gradient | get model time | report gradient time | lookup/update embedding time in report gradient
--- | --- | --- | --- | --- | --- | ---
tf.keras.layers.Embedding |30.5|3.1 |0 | 7.9 | 6.9 | 0
elasticdl.layers.Embedding with Redis |181.8 |110.7 |99.3 |6.4 |62.3 | 56.6
elasticdl.layers.Embedding with grpc | 39.8 |23.6| 3.4 | 3.3 | 7.5 | 1.8

`elasticdl.layers.Embedding with Redis` uses Redis to store the embedding tables.  Both workers and the master access Redis.
`elasticdl.layers.Embedding with grpc` store the embedding table in master. Workers lookup embeddings use grpc call. The master updates the embedding tables locally.

Running script is :

```
python3 -m elasticdl.python.elasticdl.client train \
  --image_base=elasticdl:gpudev \
  --docker_image_prefix=reg.docker.alibaba-inc.com/your-account-id/ \
  --model_def=deepfm_edl_embedding.deepfm_edl_embedding.custom_model \
  --model_zoo=model_zoo\
  --cluster_spec=path-to-sigma-file/sigma.py \
  --training_data_dir=/testdata/elasticdl/frappe/train \
  --evaluation_data_dir=/testdata/elasticdl/frappe/test \
  --evaluation_steps=0 \
  --num_epochs=1 \
  --master_resource_request="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --worker_resource_request="cpu=1,gpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --minibatch_size=256 \
  --records_per_task=1280 \
  --num_workers=2 \
  --grads_to_wait=4 \
  --job_name=test-embedding-edl \
  --log_level=INFO \
  --image_pull_policy=Always \
  --namespace=kubemaker \
  --volume='claim_name=glusterfs-csi-pv,mount_path=/testdata'
```

Replace `model_def` with `deepfm_functional_api.deepfm_functional_api.custom_model` for keras embedding layer.

**Comment by ywskycn** [comment]
@skydoorkai we can close this issue, or we need to put the results in some perf documents?

**Comment by skydoorkai** [comment]
Timing for new distributed PS @10/20/2019.

model | total time | compute gradient time | lookup embedding time in compute gradient | get model time | report gradient time | total_time - training process time\*
--- | --- | --- | --- | --- | ---|---
tf.keras.layers.Embedding with 1-ps |19.45|2.24 |0 | 4.06 | 3.63 | 9.52
elasticdl.layers.Embedding with 1-ps | 28.21 |7.31| 4.68 | 2.29 | 12.56 | 6.05
elasticdl.layers.Embedding with 2-ps | 23.27 |6.08| 3.52 | 1.59 | 8.58 | 7.02

\* `training process time` is the sum of `compute gradient time`, `get model time` and `report gradient time`.  `total_time - training process time` is spent on training data pre-processing. When the training process becomes faster, data pre-processing will become a bottleneck.

Testing script:
```
python3 -m elasticdl.python.elasticdl.client train \
  --image_base=elasticdl:gpudev \
  --docker_image_repository=your_repository \
  # for tf.keras.layers.Embedding
  --model_def=deepfm_functional_api.deepfm_functional_api.custom_model \
  # for elasticdl.layers.Embedding
  --model_def=deepfm_edl_embedding.deepfm_edl_embedding.custom_model \
  --num_ps_pods=1 or 2 \
  --use_async=False \
  --model_zoo=model_zoo\
  --cluster_spec=path_to/sigma.py \
  --training_data=/testdata/elasticdl/frappe/train \
  --num_epochs=1 \
  --master_resource_request="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --worker_resource_request="cpu=2,gpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --ps_resource_request="cpu=1,memory=2048Mi,ephemeral-storage=5000Mi" \
  --minibatch_size=256 \
  --num_minibatches_per_task=8 \
  --num_workers=2 \
  --grads_to_wait=4 \
  --job_name=test-embedding-edl \
  --log_level=INFO \
  --image_pull_policy=Always \
  --namespace=kubemaker \
  --volume='claim_name=glusterfs-csi-pv-ii,mount_path=/testdata'
```
* Distributed PS is much faster than the Redis-based system (running time reduced from 181.8 to 23.27).
* 2-PS is faster than 1-PS (23.27 vs 28.21) because of distributed communication and computation.
* As we have improved GRPC service message processing, the Keras-based embedding model is also faster than the corresponding test on  8/27/2019.
* When training process becomes faster, data pre-processing will become a bottleneck. We need to improve RecordIO-based data precessing pipeline. Currently, there are too many deserialization/copy passes.
